src/lakefs_spec/spec.py

```diff
@@ -521,6 +521,15 @@ def ls(
                         }
                     )
                 elif isinstance(obj, ObjectInfo):
+                    # Skip over prefix-only matches, e.g.:
+                    # - foo/bar
+                    # -
+                    # - foo/bar_test.txt
+
+                    # `ls("foo/bar")` should not include `bar_test.txt`
+                    if not (prefix == "" or prefix.endswith("/")) and obj.path != prefix:
```

Comment:
Note to self: This is equivalent to `if prefix and not prefix.endswith("/") and obj.path != prefix`.

Condition 1 is the repo root (which is always a directory), condition two is an explicit directory ls request, and condition 3 is a file stat request (like from `info()`).

So the condition can be thought of as "if the requested prefix does not look like a directory, and it does not equal the object path, we assume it's a file matching the prefix and ignore it", which is what we want.

Note that in the case of a directory without the terminating slash, we retry below (L547+) with the slash-terminated path.